### ISSUE TITLE
Project future

### ISSUE THREAD

**Issue by dcollinsf5** [instructure/ftl-engine#6]
Hello, and thanks for sharing your code!  I'm working on a project that we are considering using the ftl-engine project for, and really like a lot of things about it.  We are curious about what your plans are for the future of the project; do you intend to continue development or was the intent to open source the code as it is now and move on to other things?

**Comment by addisonj** [comment]
Oh hi,

didn't expect to see anyone looking at this yet, especially with the lack of docs or README ;)

This is still definitely in active development, we are also using it in production for a number of workflows. We are slowly polishing things / adding a few more features (mainly a UI for monitoring workflows) before we were planning on doing a bigger release/announcement, but the biggest gap at this point is a lack of docs.

In other words, its something we will be using for the foreseeable future and hope to make it a  vibrant community project, but it has been a bit slow getting the time to polish it and make it more accessible.

That said, if you want to dig in and if I can help in anyway, please let me know! If there is interest, I will prioritize getting some basic docs and quick start guide going.

**Comment by dcollinsf5** [comment]
Thanks for the response; glad to hear it...  I think ftl-engine and simple-swf are great  :).

I've actually been looking at the projects for a few weeks now, and spent a good few hours sifting through both ftl-engine and simple-swf trying to get a clear picture of how everything works.  Although it's been fun without docs, (and enlightening, as I haven't used typescript before) I admit there are a few spots that I feel like I don't fully understand the rationale behind, particularly around the task graph including static parameters hard-coded into the graph for every node, and the passing of a **complete** env object to every task during a workflow execution (env gets huge in our workflows and is almost always claimchecked for every task after a while even though most only care about a tiny subset of the env).

We are defining complex, multi-layer workflows with many steps that require the result of previous tasks performed in the graph to perform others further down.  We almost never have all of the parameters for every activity for a full workflow execution in scope when we generate the task graph, which is problematic for the  static validateTask() approach; we generally need to perform run-time validation of input parameters (env), since we won't know them at graph creation time.  To this end, we are transforming the 'env' that a child task or workflow execution receives at runtime.

It seems like using env is the way to pass run-time data between tasks in a workflow execution, but using env exclusively results in naming collision problems very quickly with more complex workflows; we need to control where the result of a particular task execution is put on the env object from the workflow scope.  If we need to run multiple executions of the same activity or workflow in a workflow execution, but for different purposes (e.g. we have an activity 'updateRecord', and we want to run it 2-3 times in the same workflow execution but for different reasons/targets), we want to ensure that the results of the first execution won't be overwritten by subsequent runs of the activity/workflow.  We've handled this by defining 'output' methods in the task definition object at the workflow level. Is there some trick I'm missing about using the results of one task in the next task?

I've made a clone of the repo, and implemented a rough WorkflowRegistry.  On startup, Config loads our workflow definitions, so that they can later be found by activity and decider workers.

We also are defining workflows and their dependencies as POJOs rather than keeping each workflow task as a separate file and then processing the directory structure.  We also allow nested workflows to support full dependencies.  To this end, I have created a Processor that receives a POJO where each of its properties are task definitions, rather than using the filesystem/directory processor with taskCreators, which was configured to always serialize dependencies for tasks in nested directories.  

Our workflow definitions can define (optionally) both an input and an output method so that the content of 'env' can be transformed when provided to sub-tasks and when returned back from a sub-task (either workflow or activity) - and dynamic 'parameters' (filtered env) can be validated (we are using Joi schemas) on the deciderWorker before dispatching a task.

**Comment by addisonj** [comment]
This sounds really interesting! Sounds like you have made a lot of progress already!

I am not totally surprised you ran into some limitations given your described use case. The idea for this came about as we were trying to come up with a generic way of composing a big graph of ETL tasks where execution ordered mattered, but all the actual input/output lives somewhere else (S3/HDFS for example), which basically means we can know up front the work to be done pretty easily and just are shoving a dataset through the this "pipeline". 

So while the general abstraction of using a graph to represent the work to be done has been really useful, there probably wasn't as much work put into things outside our primary use case, but we have plans for a lot that, which I am going to open a number of issues to define (at least our idea) of a roadmap.

It really seems like you have gotten the gist of things though, especially with overriding the Processor for extra behavior in graph generation and glad it was at least somewhat approachable from a code perspective!

I am going to close this for now, will open an issue to represent the work here of getting a README started
